### PR TITLE
Add validation to CompetitiveEvent dtos

### PR DESCRIPTION
**Pull request by HlibBondarev**
Add validation to CompetitiveEvent dtos (CompetitiveEventBaseDto, CompetitiveEventDto and CompetitiveEventV2Dto):

- added attributes to CompetitiveEventBaseDto properties;
- implemented IValidatableObject interface for CompetitiveEventBaseDto, CompetitiveEventDto and CompetitiveEventV2Dto classes;
- added validation for ImageFiles and ImageIds properties in the classes - CompetitiveEventV2Dto and CompetitiveEventV2CreateRequestDto;
- removed CompetitiveEventCreateUpdateDto; 
- used CompetitiveEventBaseDto instead in the classes - CompetitiveEventV2CreateRequestDto, CompetitiveEventService, CompetitiveEventController and the interface - ICompetitiveEventService; 
- improved the Create() action in the CompetitiveEvent and CompetitiveEventDraft controllers — added validation of the ImageFiles and ImageIds properties;
- improved the Approve() method of the CompetitiveEventDraftService class - added null checking for the CompetitiveEvent draft;
- added appropriate tests.

Competitive event Create-Update

## Summary by CodeRabbit

* **New Features**
  * Stronger validation for competitive events (cross-field checks for times, seats, ages) and stricter image/cover requirements for V2 create/draft flows.

* **Bug Fixes**
  * Draft approval now errors when a draft is missing; controllers return BadRequest for invalid image inputs and NoContent when an item is not found.

* **Refactor**
  * Unified create/update DTO surface and changed default workshop pay-rate to "None".

* **Tests**
  * Added/updated validation, controller, and service tests for new rules, image uploads, transactional draft flows, and edge cases.

### CODE REVIEW

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/CompetitiveEventBaseDto.cs [hunk 1]
```diff
@@ -88,4 +107,122 @@ public class CompetitiveEventBaseDto: IHasContactsDto<OutOfSchool.Services.Model
...
+        AreThereBenefits = dto.AreThereBenefits ?? false,
+        Benefits = dto.Benefits,
+        MinimumAge = dto.MinimumAge,
+        MaximumAge = dto.MaximumAge ?? 0,
+        Price = dto.Price ?? 0,
+        CompetitiveSelection = dto.CompetitiveSelection ?? false,
+        Contacts = dto.Contacts?.ToModel(),
+        SubDirections = dto.SubDirectionIds.Select(id => new SubDirection { DirectionId = id }).ToList(),
+        CoverImageId = dto.CoverageId.ToString(),
+    };
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/CompetitiveEventBaseDto.cs]: Ok

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/V2/CompetitiveEventV2CreateRequestDto.cs [hunk 2]
```diff
@@ -18,6 +18,27 @@ public class CompetitiveEventV2CreateRequestDto : CompetitiveEventCreateUpdateDt
...
+
+        if ((CoverImage is null) == string.IsNullOrEmpty(CoverImageId))
+        {
+            yield return new ValidationResult(
+                "Must be filled either CoverImage or CoverImageId, but not both",
+                [nameof(CoverImage), nameof(CoverImageId)]);
+        }
+
+        if ((ImageFiles ?? []).Count == 0 && (ImageIds ?? []).Count == 0)
+        {
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/V2/CompetitiveEventV2CreateRequestDto.cs:36]: No, I've changed test - CompetitiveEventsV2ControllerTests.Create_ReturnsCreated_WhenSuccessful

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/TempSave/WorkshopRequiredPropertiesDto.cs [hunk 3]
```diff
@@ -40,7 +40,7 @@ public class WorkshopRequiredPropertiesDto : WorkshopMainRequiredPropertiesDto
 
     [EnumDataType(typeof(PayRateType), ErrorMessage = Constants.EnumErrorMessage)]
     [RequiredIf(nameof(IsPaid), true, ErrorMessage = "PayRate is required")]
-    public PayRateType? PayRate { get; set; } = PayRateType.Class;
+    public PayRateType? PayRate { get; set; } = PayRateType.None;
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/TempSave/WorkshopRequiredPropertiesDto.cs:43]: ok, I've deleted attribute - [RequiredIf(nameof(IsPaid), true, ErrorMessage = "PayRate is required")]

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs [hunk 4]
```diff
@@ -127,6 +127,8 @@ public async Task<IActionResult> Create([FromForm] CompetitiveEventV2CreateReque
             null => BadRequest("CompetitiveEvent is null"),
             _ when !ModelState.IsValid => BadRequest(ModelState),
             _ when await IsCurrentUserBlocked() => StatusCode(403, "User is blocked"),
+            _ when (dto.ImageFiles ?? []).Count == 0 || (dto.ImageIds ?? []).Count > 0 
+            => BadRequest("When creating CompetitiveEvent, the ImageFiles field must contain a non-empty array of images, and the ImageIds field must be null or an empty array."),
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventController.cs:131]: No.
CompetitiveEventV2CreateRequestDto used for both operation - Create and Update.
Rule ((dto.ImageFiles ?? []).Count == 0 || (dto.ImageIds ?? []).Count > 0) acts only for Create, but for Update acts rule checked in CompetitiveEventV2CreateRequestDto.Validate.

File: OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventDraftController.cs [hunk 5]
```diff
@@ -56,35 +56,50 @@ public CompetitiveEventDraftController(
...
+        var error = dto switch
         {
-            return BadRequest(ModelState);
+            null => BadRequest("CompetitiveEvent is null"),
+            _ when !ModelState.IsValid => BadRequest(ModelState),
+            _ when (dto.ImageFiles ?? []).Count == 0 || (dto.ImageIds ?? []).Count > 0
+            => BadRequest("When creating a draft of CompetitiveEvent, the ImageFiles field must contain a non-empty array of images, and the ImageIds field must be null or an empty array."),
+            _ => null
+        };
+
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.WebApi/Controllers/V2/CompetitiveEventDraftController.cs:70]: The explanation is similar - https://github.com/ita-social-projects/OoS-Backend/pull/1978#discussion_r2387060157

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/CompetitiveEventsV2ControllerTests.cs [hunk 6]
```diff
@@ -100,7 +101,11 @@ public async Task GetByProviderId_ReturnsNoContent_WhenNoResultsFound()
     [Test]
     public async Task Create_ReturnsCreated_WhenSuccessful()
     {
-        var dto = new CompetitiveEventV2CreateRequestDto();
+        var dto = new CompetitiveEventV2CreateRequestDto() 
+        { 
+            ImageFiles = [It.IsAny<IFormFile>()],
+            CoverImage = It.IsAny<IFormFile>()        
+        };
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/CompetitiveEventsV2ControllerTests.cs]: Ok

File: OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/CompetitiveEventsV2ControllerTests.cs [hunk 7]
```diff
@@ -147,7 +152,7 @@ public async Task Create_ReturnsForbidden_WhenUserIsBlocked()
     [Test]
     public async Task Create_ReturnsBadRequest_WhenServiceThrowsInvalidOperation()
     {
-        var dto = new CompetitiveEventV2CreateRequestDto();
+        var dto = new CompetitiveEventV2CreateRequestDto() { ImageFiles = [It.IsAny<IFormFile>()] };
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Controllers/CompetitiveEventsV2ControllerTests.cs]: Ok

File: OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/CompetitiveEventBaseDto.cs [hunk 8]
```diff
@@ -88,4 +107,120 @@ public class CompetitiveEventBaseDto: IHasContactsDto<OutOfSchool.Services.Model
...
+        TermsOfParticipation = dto.TermsOfParticipation,
+        AreThereBenefits = dto.AreThereBenefits ?? false,
+        Benefits = dto.Benefits,
+        MinimumAge = dto.MinimumAge,
+        MaximumAge = dto.MaximumAge ?? 0,
+        Price = dto.Price ?? 0,
+        CompetitiveSelection = dto.CompetitiveSelection ?? false,
+        Contacts = dto.Contacts?.ToModel(),
+        SubDirections = [.. (dto.SubDirectionIds ?? []).Select(id => new SubDirection { Id = id })],
+    };
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.BusinessLogic/Models/CompetitiveEvent/CompetitiveEventBaseDto.cs:185]: no

File: OutOfSchool/OutOfSchool.WebApi.Tests/Services/CompetitiveEventsV2ServiceTests.cs [hunk 9]
```diff
@@ -214,20 +214,26 @@ public async Task UpdateV2_ReturnResult_WhenEntityExists()
         };
 
         repoMock.Setup(r => r.GetByIdWithDetails(dto.Id, It.IsAny<string>(), It.IsAny<Func<IQueryable<CompetitiveEvent>, IQueryable<CompetitiveEvent>>>()))
-            .ReturnsAsync(competitiveEvent).Verifiable(Times.Once); ;
+            .ReturnsAsync(competitiveEvent)
+            .Verifiable(Times.Once); ;
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Services/CompetitiveEventsV2ServiceTests.cs]: ok

File: OutOfSchool/OutOfSchool.WebApi.Tests/Services/CompetitiveEventsV2ServiceTests.cs [hunk 10]
```diff
@@ -214,20 +214,26 @@ public async Task UpdateV2_ReturnResult_WhenEntityExists()
...
         repoMock.Setup(r => r.RunInTransaction(It.IsAny<Func<Task<(CompetitiveEvent, MultipleImageChangingResult, ImageChangingResult)>>>()))
-            .Returns<Func<Task<(CompetitiveEvent, MultipleImageChangingResult, ImageChangingResult)>>>(f => f()).Verifiable(Times.Once);;
+            .Returns<Func<Task<(CompetitiveEvent, MultipleImageChangingResult, ImageChangingResult)>>>(f => f())
+            .Verifiable(Times.Once);;
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.WebApi.Tests/Services/CompetitiveEventsV2ServiceTests.cs]: Ok

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs [hunk 11]
```diff
@@ -68,29 +68,39 @@ await currentUserService.UserHasRights(new ProviderRights(existingCompetitiveEve
...
 
         return new CompetitiveEventDraftResultDto
         {
-            CompetitiveEventDraft = await MapCompetitiveEventDraftWithDetails(createdCompetitiveEventDraft),
-            UploadingCoverImagesCompetitiveEventResult = uploadImagesResult.UploadingCoverImageResult,
-            UploadingImagesResults = uploadImagesResult.UploadingImagesResults?.MultipleKeyValueOperationResult
+            CompetitiveEventDraft = await MapCompetitiveEventDraftWithDetails(draftImageUpdateResult.Value.createdCompetitiveEventDraft),
+            UploadingCoverImagesCompetitiveEventResult = draftImageUpdateResult.Value.uploadImagesResult.UploadingCoverImageResult,
+            UploadingImagesResults = draftImageUpdateResult.Value.uploadImagesResult.UploadingImagesResults.MultipleKeyValueOperationResult
         };
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs:111]: Ok

File: OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs [hunk 12]
```diff
@@ -312,6 +322,11 @@ public async Task Approve(Guid id)
 
         var competitiveEventDraft = await GetDraftById(id);
 
+        if (competitiveEventDraft == null)
+        {
+            throw new ArgumentException($"There is no CompetitiveEvent draft with such Id.");
+        }
```
[the remark this answers is not in the document]
Comment (HlibBondarev) [reply OutOfSchool/OutOfSchool.BusinessLogic/Services/CompetitiveEventDrafts/CompetitiveEventDraftService.cs]: Ok

**Review by DmyMi — APPROVED** [review]
